services/worker-manager/src/providers/aws.js

```diff
@@ -0,0 +1,306 @@
+        InstanceIds: [worker.workerId.toString()],
+        IncludeAllInstances: true,
+      }).promise()).InstanceStatuses;
+    } catch (e) {
+      if (e.code === 'InvalidInstanceID.NotFound') { // aws throws this error for instances that had been terminated, too
+        return worker.modify(w => {w.state = this.Worker.states.STOPPED;});
+      }
+      throw e;
+    }
+
+    Promise.all(instanceStatuses.map(is => {
+      switch (is.InstanceState.Name) {
+        case 'pending':
+        case 'running':
+        case 'shutting-down': //so that we don't turn on new instances until they're entirely gone
+        case 'stopping':
+          this.seen[worker.workerPoolId] += 1;
+          return Promise.resolve();
+
+        case 'terminated':
+        case 'stopped':
+          return worker.modify(w => {w.state = this.Worker.states.STOPPED;});
+
+        default:
+          return Promise.reject(`Unknown state: ${is.InstanceState.Name} for ${is.InstanceId}`);
+      }
+    }));
+  }
+
+  // should this be implemented on Provider? Looks like it's going to be the same for all providers
+  async scanPrepare() {
+    this.seen = {};
+    this.errors = {};
+  }
+
+  async scanCleanup() {
+    await Promise.all(Object.entries(this.seen).map(async ([workerPoolId, seen]) => {
+      const workerPool = await this.WorkerPool.load({
+        workerPoolId,
+      }, true);
+
+      if (!workerPool) {
+        return; // In this case, the worker pool has been deleted so we can just move on
+      }
+
+      await workerPool.modify(wp => {
+        if (!wp.providerData[this.providerId]) {
+          wp.providerData[this.providerId] = {};
+        }
+        wp.providerData[this.providerId].running = seen;
+      });
+    }));
+  }
+
+  /**
+   * Method to select instance launch specification. At the moment selects at random, which is temporary
+   *
+   * @param possibleConfigs Array<Object>
+   * @returns Object
+   */
+  chooseConfig({possibleConfigs}) {
+    assert(Array.isArray(possibleConfigs), 'possible configs is not an array');
+
+    const i = Math.floor(Math.random() * possibleConfigs.length);
+
+    return possibleConfigs[i];
+  }
+
+  /**
+   * Method to verify the instance identity document against the signature and public key
+   * The signature is the one that is obtained by calling
+   * http://169.254.169.254/latest/dynamic/instance-identity/signature
+   * endpoint. The endpoint produces base64-encoded data, so no conversions
+   * are necessary prior to sending
+   *
+   * @param document - a string or a buffer
+   * @param signature - base64-encoded data (can be a string or buffer)
+   * @returns boolean (true if verification is successful)
+   */
+  verifyInstanceIdentityDocument({document, signature}) {
+    const verifier = crypto.createVerify('sha256');
+
+    verifier.update(document);
+    let result;
+
+    result = verifier.verify(this.ec2iid_RSA_key.toString(), signature.toString(), 'base64');
+
+    return result;
```

Comment:
Could be simplified to just `return verifier.verify(..)`.

Reply:
oh that var was for debugging. thank you for catching that!